crates/matrix-sdk-ui/src/timeline/controller/observable_items.rs

```diff
@@ -478,3 +478,286 @@ impl AllRemoteEvents {
         }
     }
 }
+
+#[cfg(test)]
+mod all_remote_events_tests {
+    use assert_matches::assert_matches;
+    use ruma::event_id;
+
+    use super::{AllRemoteEvents, EventMeta};
+
+    fn event_meta(event_id: &str, timeline_item_index: Option<usize>) -> EventMeta {
+        EventMeta { event_id: event_id.parse().unwrap(), timeline_item_index, visible: false }
+    }
+
+    macro_rules! assert_events {
+        ( $events:ident, [ $( ( $event_id:literal, $timeline_item_index:expr ) ),* $(,)? ] ) => {
+            let mut iter = $events .iter();
+
+            $(
+                assert_matches!(iter.next(), Some(EventMeta { event_id, timeline_item_index, .. }) => {
+                    assert_eq!(event_id.as_str(), $event_id );
+                    assert_eq!(*timeline_item_index, $timeline_item_index );
+                });
+            )*
+
+            assert!(iter.next().is_none(), "Not all events have been asserted");
+        }
+    }
+
+    #[test]
+    fn test_clear() {
+        let mut events = AllRemoteEvents::default();
+
+        events.push_back(event_meta("$ev0", None));
+        events.push_back(event_meta("$ev1", None));
+        events.push_back(event_meta("$ev2", None));
+
+        assert_eq!(events.iter().count(), 3);
+
+        events.clear();
+
+        assert_eq!(events.iter().count(), 0);
+    }
+
+    #[test]
+    fn test_push_front() {
```

Comment:
Not going to require it, but something to keep in mind, these container types and insertions/removals sound like they would be good candidates for property based tests.

Reply:
Yes, good idea. Let's address that in a future PR. This one is already large enough.